Lib/http/server.py

```diff
@@ -687,25 +687,30 @@ def send_head(self):
         except OSError:
             self.send_error(HTTPStatus.NOT_FOUND, "File not found")
             return None
+
+        fs = os.fstat(f.fileno())
+            
         try:
             # Use browser cache if possible
-            if "If-Modified-Since" in self.headers:
+            if "If-Modified-Since" in self.headers \
+                    and not "If-None-Match" in self.headers:
                 # compare If-Modified-Since and date of last file modification
-                fs = os.stat(path)
                 ims = email.utils.parsedate(self.headers["If-Modified-Since"])
```

Comment:
Ok, the latest commit uses `parsedate_to_datetime()` and checks that it has a timezone set to UTC, which should be the case for the header (cf [RFC 7231](https://tools.ietf.org/html/rfc7231#section-7.1.1.1)). To compare with last modification time, both must be timezone-aware datetimes.